File: code/sort_algorithms.hs
import Data.List (insert, unfoldr, transpose)


-- bubble sort
bubbleSort :: (Ord a) => [a] -> [a]
bubbleSort [] = []
bubbleSort [x] = [x]
bubbleSort x = 
  (iterate swapPass x) !! ((length x) - 1)
  where swapPass [x] = [x]
        swapPass (x1:x2:xs)
          | x1 > x2 = x2 : swapPass (x1:xs)
          | otherwise = x1 : swapPass (x2:xs)


-- insertion sort
insertionSort :: (Ord a) => [a] -> [a]
insertionSort [] = []
insertionSort [x] = [x]
insertionSort (x:xs) = 
  insert $ insertionSort xs
  where insert [] = [x]
        insert (y:ys)
          | x < y = x : y : ys
          | otherwise = y : insert ys


-- cocktail sort
cocktailSort :: (Ord a) => [a] -> [a]
cocktailSort [] = []
cocktailSort [x] = [x]
cocktailSort x =
  [head pass] ++ cocktailSort middle ++ [last pass]
  where pass = swapPass (<) $ reverse $ swapPass (>) $ reverse x
        middle = drop 1 $ reverse $ drop 1 $ reverse pass
        swapPass :: (a -> a -> Bool) -> [a] -> [a]
        swapPass _ [] = []
        swapPass _ [x] = [x]
        swapPass f (x1:x2:xs)
          | f x1 x2 = x1 : swapPass f (x2:xs)
          | otherwise = x2 : swapPass f (x1:xs)


-- quick sort
quickSort :: (Ord a) => [a] -> [a]
quickSort [] = []
quickSort [x] = [x]
quickSort (x:xs) =
  let smaller = quickSort [y | y <- xs, y <= x]
      bigger = quickSort [y | y <- xs, y > x]
  in smaller ++ [x] ++ bigger


-- shell sort
shellSort :: (Ord a) => [a] -> [a]
shellSort [] = []
shellSort [x] = [x]
shellSort xs =
  foldr epoch xs gaps
  where gaps = takeWhile (< length xs) $ iterate (2*) 1
        epoch gap = concat .                    -- [1, 2, 3, 4, 5]
                    transpose .                 -- [[1, 2], [3, 4], [5]]
                    map (foldr insert []) .     -- sort
                    transpose .                 -- [[1, 3, 5], [2, 4]]
                    takeWhile (not . null) .    -- [[1, 2], [3, 4], [5]]
                    unfoldr(Just . splitAt gap) -- [[1, 2], [3, 4], [5], [], [], []]

File: code/sort_algorithms.rs
fn bubble_sort(list: &Vec<i32>) -> Vec<i32> {
  let mut arr = list.to_vec();

  for i in (0 .. arr.len()).rev() {
    for j in 0 .. i {
      if arr[j] > arr[j + 1] {
        arr.swap(j, j + 1);
      }
    }
  }

  arr
}


fn insertion_sort(list: &Vec<i32>) -> Vec<i32> {
  let mut arr = list.to_vec();

  for i in 1 .. arr.len() {
    let mut j = i;

    while j > 0 && arr[j - 1] > arr[j] {
      arr.swap(j - 1, j);
      j -= 1;
    }
  }

  arr
}


fn cocktail_sort(list: &Vec<i32>) -> Vec<i32> {
  let mut arr = list.to_vec();
  let mut left = 0;
  let mut right = arr.len() - 1;

  while left < right {
    for i in left .. right {
      if arr[i] > arr[i + 1] {
        arr.swap(i, i + 1);
      }
    }

    for i in (left + 1 .. right).rev() {
      if arr[i] < arr[i - 1] {
        arr.swap(i - 1, i);
      }
    }

    left += 1;
    right -= 1;
  }

  arr
}

fn sort_helper(arr: &mut Vec<i32>, left: isize, right: isize) {
  if left >= right {
    return
  }

  let m = (left + right) / 2;
  let pivot = arr[m as usize];
  let mut i = left;
  let mut j = right;

  while i < j {
    while arr[i as usize] < pivot {
      i += 1;
    }

    while arr[j as usize] > pivot {
      j -= 1;
    }


    if i <= j {
      arr.swap(i as usize, j as usize);
      i += 1;
      j -= 1;
    }
  }

  if left < j {
    sort_helper(arr, left, j);
  }

  if i < right {
    sort_helper(arr, i, right);
  }
}

fn quick_sort(list: &Vec<i32>) -> Vec<i32> {
  let mut arr = list.to_vec();
  let n = arr.len();
  sort_helper(&mut arr, 0, n as isize - 1);
  arr
}


fn main () {
  let arr = vec![1, 6, 2, -1, 3, 0, 1];
  let bubble_sorted = bubble_sort(&arr);
  let insert_sorted = insertion_sort(&arr);
  let shaker_sorted = cocktail_sort(&arr);
  let quick_sorted = quick_sort(&arr);
  println!("Bubbl sort: {:?}", bubble_sorted);
  println!("Insertion sort: {:?}", insert_sorted);
  println!("Cocktail sort: {:?}", shaker_sorted);
  println!("Quick sort: {:?}", quick_sorted);
}
